gapis/api/vulkan/state_rebuilder.go

```diff
@@ -0,0 +1,2582 @@
+		sb.createCommandPool(s.CommandPools.Get(cp))
+	}
+
+	for _, pc := range s.PipelineCaches.KeysSorted() {
+		sb.createPipelineCache(s.PipelineCaches.Get(pc))
+	}
+
+	for _, dsl := range s.DescriptorSetLayouts.KeysSorted() {
+		sb.createDescriptorSetLayout(s.DescriptorSetLayouts.Get(dsl))
+	}
+
+	for _, pl := range s.PipelineLayouts.KeysSorted() {
+		sb.createPipelineLayout(s.PipelineLayouts.Get(pl))
+	}
+
+	for _, rp := range s.RenderPasses.KeysSorted() {
+		sb.createRenderPass(s.RenderPasses.Get(rp))
+	}
+
+	for _, sm := range s.ShaderModules.KeysSorted() {
+		sb.createShaderModule(s.ShaderModules.Get(sm))
+	}
+
+	for _, cp := range GetPipelinesInOrder(s, true) {
+		sb.createComputePipeline(s.ComputePipelines.Get(cp))
+	}
+
+	for _, gp := range GetPipelinesInOrder(s, false) {
+		sb.createGraphicsPipeline(s.GraphicsPipelines.Get(gp))
+	}
+
+	for _, iv := range s.ImageViews.KeysSorted() {
+		sb.createImageView(s.ImageViews.Get(iv))
+	}
+
+	for _, bv := range s.BufferViews.KeysSorted() {
+		sb.createBufferView(s.BufferViews.Get(bv))
+	}
+
+	for _, dp := range s.DescriptorPools.KeysSorted() {
+		sb.createDescriptorPool(s.DescriptorPools.Get(dp))
+	}
+
+	for _, fb := range s.Framebuffers.KeysSorted() {
+		sb.createFramebuffer(s.Framebuffers.Get(fb))
+	}
+
+	for _, fb := range s.DescriptorSets.KeysSorted() {
+		sb.createDescriptorSet(s.DescriptorSets.Get(fb))
+	}
+
+	for _, qp := range s.QueryPools.KeysSorted() {
+		sb.createQueryPool(s.QueryPools.Get(qp))
+	}
+
+	for _, qp := range s.CommandBuffers.KeysSorted() {
+		sb.createCommandBuffer(s.CommandBuffers.Get(qp), VkCommandBufferLevel_VK_COMMAND_BUFFER_LEVEL_SECONDARY)
+	}
+
+	for _, qp := range s.CommandBuffers.KeysSorted() {
+		sb.createCommandBuffer(s.CommandBuffers.Get(qp), VkCommandBufferLevel_VK_COMMAND_BUFFER_LEVEL_PRIMARY)
+	}
+
+	return sb.cmds
+}
+
+func GetPipelinesInOrder(s *State, compute bool) []VkPipeline {
+
+	pipelines := []VkPipeline{}
+	unhandledPipelines := map[VkPipeline]VkPipeline{}
+	handledPipelines := map[VkPipeline]bool{}
+	if compute {
+		for _, p := range s.ComputePipelines.KeysSorted() {
+			pp := s.ComputePipelines.Get(p)
+			unhandledPipelines[pp.VulkanHandle] = pp.BasePipeline
+		}
+	} else {
+		for _, p := range s.GraphicsPipelines.KeysSorted() {
+			pp := s.GraphicsPipelines.Get(p)
+			unhandledPipelines[pp.VulkanHandle] = pp.BasePipeline
+		}
+	}
+
+	numHandled := 0
+	for len(unhandledPipelines) != 0 {
+		for k, v := range unhandledPipelines {
+			handled := false
+			if v == VkPipeline(0) {
+				pipelines = append(pipelines, k)
+				handled = true
+			} else if _, ok := handledPipelines[v]; ok {
+				pipelines = append(pipelines, k)
+				handled = true
+			}
+			if handled {
+				handledPipelines[k] = true
+				delete(unhandledPipelines, k)
+				numHandled++
+			}
+		}
+		if numHandled == 0 {
+			// There is a cycle in the basePipeline indices.
+			// Or the no base pipelines does exist.
+			// Create the rest without base pipelines
```

Comment:
Is it allowed to by cyclical? Probably we should emit an error in the report view for this?

Reply:
I am not sure if it is technically allowed, but I want to at least continue to work if it is.
Hopefully the validation layers can catch this.